### PR TITLE
Use systemd Boot Assessment

### LINKED ISSUE #8: Option to disable health-checker in Grub
Hello,

I would like to be able to do the following:

force disable the health-checker via a grub flag.
My reasoning is the that currently selinux is in a failed state in MicroOS.

But what happens if I check selinux with the health-checker? 
An endless cycle of reboots. (theoretically, I did not try it)

From a security perspective I may want to check the state of selinux during start up.
Let me disable the health-check for debugging purposes if needed.

### PR DESCRIPTION
**Pull request by danyspin97**
[Automatic Boot Assessment](https://systemd.io/AUTOMATIC_BOOT_ASSESSMENT/) allows systemd-boot and systemd to mark boot entries as either good or bad, depending on if they can boot successfully or not.

This PR changes health-checker into a service that is part of the automatic boot assessment, by using the special target `boot-complete.target`. When systemd-boot and `/etc/kernel/tries` is greater than 0, the current boot entry get renamed to start the counting.

If health-checker tests pass without errors, then the boot entry is marked as good by systemd-bless-boot. If there is any error, then health-checker decides if there should be a reboot, or it should start an emergency shell. If the current entry is the default one and the entry has still some tries left (i.e. it has not been marked as `bad`), then reboot. If the current entry is the default one and there are no tries left, then start an emergency shell. The default entry will be picked among the one that are known to work or haven't been tested yet, so the emergency shell is only started when all entries have been tried (this could lead to many reboots). If the user choose an entry instead of letting systemd-boot pick the default one, then health-checker will not reboot by default (this can be enforced with the argument below).

I have also added two kernel cmdline arguments to fix #8 :
- `health-checker-reboot`:
  - `force`: always reboot when health-checker fails and the loaded boot entry is not the default one
  -  `disable`: health-checker never reboots
- `health-checker=disabled`: skill all tests and mark health-checker as successful. This breaks systemd Automatic Boot Counting but helps with debugging or some edge cases.

## Requirements
- `/etc/kernel/tries` to have a number greater than 0. Currently, I am shipping this file in the health-checker package.

## Current blockers
- systemd-bless-boot cannot rename the boot entries, due to selinux enforcing policy. Bug tracked [here](https://bugzilla.suse.com/show_bug.cgi?id=1232527).

### CODE REVIEW

File: sbin/health-checker.in [hunk 1]
```diff
@@ -10,24 +10,12 @@
 # - if reboot does not help, stop system before further damage is done
 #
 
-STATE_FILE=/var/lib/misc/health-check.state
-REBOOTED_STATE=/var/lib/misc/health-check.rebooted
 PLUGINDIR=@PLUGINDIR@
 USR_LOCAL_PLUGINDIR=/usr/local/libexec/health-checker
+ETC_PLUGINDIR=/etc/health-checker
```
Comment (thkukuk) [inline sbin/health-checker.in]: Please, don't add scripts to /etc, the mid- to long term plan is still to make /etc non-executable. Only remaining problem is grub...
And why do we need /etc/health-checker?
If somebody wants to provide plugins via RPM, /usr/libexec/health-checker is the correct location.
If somebody wants to use something else, /usr/local/libexec/health-checker is the correct place.
I don't see a need for /etc/health-checker.

File: systemd/health-checker.service [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
-After=etcd.service
-After=kubelet.service
 After=rebootmgr.service
 After=systemd-logind.service
 Wants=local-fs.target
+Before=boot-complete.target
 
 [Service]
 Type=oneshot
+RemainAfterExit=yes
```
Comment (thkukuk) [inline systemd/health-checker.service]: That's a duplicate, see two lines below.

**Review by thkukuk — CHANGES_REQUESTED** [review]

File: systemd/health-checker.service (hunk 2 above)
Comment (danyspin97) [reply systemd/health-checker.service]: I see, this has been added in the main branch after my changes. I have removed it.

File: sbin/health-checker.in (hunk 1 above)
Comment (danyspin97) [reply sbin/health-checker.in]: Thanks for the explanation Thorsten. `/etc/health-checker` has been added just for testing purposes because the virtual machines I am generating on OBS have some issues with `/usr/local/libexec/health-checker`. They don't work with `/etc/health-checker` regardless, so I'll remove the new path and try the official images. I should have added a comment regarding the testing, my bad.

**Comment by Vogtinator** [comment]
What do we do on platforms without EFI vars? Just declare them unsupported by this mechanism?

**Comment by danyspin97** [comment]
> What do we do on platforms without EFI vars? Just declare them unsupported by this mechanism?

This version uses EFI vars for simplicity, since it makes it easier to retrieve the current and default boot entries. This can work with any bootloader as long as we know this info and Automatic Boot Assessment is supported.

**Comment by danyspin97** [comment]
I replaced the EFI variables by reimplementing Automatic Boot Assessment logic in health-checker. I calculate the default one as being the first entry, descending order based on the name, that also has not been disabled by the boot counting. It is quite bare bones, but it works. One possible issue would be having different kernel versions, then health-checker would require a more robust parser. For detecting the current entry, I am checking the snapshot version of the current mounted snapshot.

**Comment by danyspin97** [comment]
I thought a little bit more on the approach I have taken in this PR. `bootctl` can set the default entry by changing the EFI variables, so I'd go back to reading the EFI variables first. I still think the bash implementation of the BLS logic for choosing the default is good as fallback for systems that don't support EFI vars.

**Comment by aplanas** [comment]
`bootctl` follows the BLS and also the BLI, that describes the set of EFI variables that the bootloader will follow. Because grub2-bls does not really follow this last BLI specification, `sdbootutil` needed to re-implement `set/get-default` and `set/get-timeout` for BLI and non-BLI bootloaders.

For example, if we are using `systemd-boot` in an architecture that does not has EFI variables, it will set the configuration in the loader.conf file in the ESP, and if we are in a `grub2-bls` system, then will set the grubenv and the EFI variable (or loader.conf), so bootctl information will always read the correct information for both bootloaders.

My recommendation is to follow this path, or use `sdbootutil set-default` and `get-default` to abstract this part.

**Comment by thkukuk** [comment]
Please adjust the README.md and manual page, especially with the now missing state file and the new kernel commandline options.
Else looks good to me.

File: README.md [hunk 3]
```diff
@@ -2,34 +2,47 @@
 
 Check the state of a openSUSE MicroOS system after a reboot.
 
+## Configure
+
+All services that should be checked, need to be listed in the 'After' section.
```
Comment (Vogtinator) [inline README.md]:
```suggestion
All services that should be checked need to be listed in the 'After' section.
```

File: README.md [hunk 4]
```diff
@@ -2,34 +2,47 @@
...
+The `health-checker` script will call all plugins inside
+`/usr/local/libexec/health-checker` directory. Every plugin is responsible to
```
Comment (Vogtinator) [inline README.md]: Should this really be `/usr/local`? That's global and not per-snapshot

File: README.md [hunk 5]
```diff
@@ -2,34 +2,47 @@
...
+check a special service or condition. For this, the plugin is called with the
+option *check*. If this fails, the plugin will exit with the return value `1`,
+else `0`. Have a look at the default plugins shipped in
+`/usr/local/libexec/health-checker` for examples.
```
Comment (Vogtinator) [inline README.md]: The `health-checker` package must not ship anything in `/usr/local`, it's not snapshotted.

File: README.md [hunk 6]
```diff
@@ -2,34 +2,47 @@
...
-   the problem, all plugins will be called with the option *stop*. At the
-   end, the machine should still run, so that an admin can login, but no
-   service should run, so that nothing can break.
+`health-checker` assess the status of the system at the boot and act accordingly.
+It uses the [Automatic Boot Assessment](https://systemd.io/AUTOMATIC_BOOT_ASSESSMENT/)
+provided by systemd to check every new snapshot, to reboot a predetermined amount
+of times, and then it lets the user access an emergency shell when everything
+else fails.
+
+Every new snapshot has a separate boot entry with a boot counter (according to
```
Comment (Vogtinator) [inline README.md:35]: "Every new snapshot" or just "Every snapshot"?

File: README.md [hunk 7]
```diff
@@ -2,34 +2,47 @@
...
+`/etc/kernel/tries`, which defaults to 3); when that snapshot will be booted
```
Comment (Vogtinator) [inline README.md]:
```suggestion
`/etc/kernel/tries`, which defaults to 3); when that snapshot is booted
```

File: README.md [hunk 8]
```diff
@@ -2,34 +2,47 @@
...
+for the first time, the bootloader (systemd-boot by default on MicroOS, but
+grub2-bls is also supported) will decrease the amount of tries left. Then if
+health-checker succeed, systemd will call `systemd-bless-boot`, which will mark
```
Comment (Vogtinator) [inline README.md]:
```suggestion
health-checker succeeded, `systemd-bless-boot.service` runs
```

File: README.md [hunk 9]
```diff
@@ -2,34 +2,47 @@
...
+the new snapshot as working. If instead `health-checker` fails and the entry is
+the default one, the system gets automatically rebooted, until the amount of
+tries equals 0. At that point an emergency shell gets opened.
```
Comment (Vogtinator) [inline README.md]: Will it not reboot into the last working snapshot?

File: sbin/health-checker.in [hunk 10]
```diff
@@ -71,77 +58,77 @@ telem_send_record()
...
-{
-    . ${STATE_FILE}
-    mount -o remount,rw /.snapshots
-    btrfs subvolume set-default ${LAST_WORKING_BTRFS_ID} /.snapshots
-    if [ $? -ne 0 ]; then
-        create_log user.crit "ERROR: btrfs set-default $BTRFS_ID_DEFAULT failed!"
-	telem_send_payload 1
-        exit 1
+systemd-bless-boot() {
+    if [ -x /usr/lib/systemd/systemd-bless-boot ]; then
```
Comment (Vogtinator) [inline sbin/health-checker.in:177]: Can this happen resp. is it supposed to happen that this does not exist? If yes, do the callers handle it properly that nothing happens?

File: sbin/health-checker.in [hunk 11]
```diff
@@ -71,77 +58,77 @@ telem_send_record()
...
-    if [ ! -f ${STATE_FILE} ]; then
-	# No state file, no successful boot
-	create_log user.emerg "Machine didn't come up correctly, stopping services"
-	stop_services
-	return
+    should_reboot=0
+    # Do not reboot by default if the entry has been chosen manually or the reboot has
+    # been disabled in the kernel cmdline
+    # selected_entry contains the boot count, remove it before comparing it to the default entry
+    if grep "health-checker-reboot=force" /proc/cmdline || ( [ "$default_entry" == "$selected_entry" ] && \
```
Comment (Vogtinator) [inline sbin/health-checker.in]:
```suggestion
    if grep -qw "health-checker-reboot=force" /proc/cmdline || ( [ "$default_entry" == "$selected_entry" ] && \
```

File: sbin/health-checker.in [hunk 12]
```diff
@@ -71,77 +58,77 @@ telem_send_record()
...
+        ! grep "health-checker-reboot=disabled" /proc/cmdline ); then
```
Comment (Vogtinator) [inline sbin/health-checker.in]:
```suggestion
        ! grep -qw "health-checker-reboot=disabled" /proc/cmdline ); then
```

File: sbin/health-checker.in [hunk 13]
```diff
@@ -71,77 +58,77 @@ telem_send_record()
...
+        # good should never appear here because systemd-bless-boot.service that
+        # marks an entry as "good" is called after boot.entry (and health-checker as well)
+        # All the next reboots for the same entry will have a "clean" state
+    esac
 }
 
-# Clear GRUB flag (used to determine if system was able to boot at all)
-echo "Clearing GRUB flag"
-grub2-editenv - set health_checker_flag=0
+if grep "health-checker=disabled" /proc/cmdline; then
```
Comment (Vogtinator) [inline sbin/health-checker.in]:
```suggestion
if grep -qw "health-checker=disabled" /proc/cmdline; then
```

File: systemd/health-checker.service [hunk 14]
```diff
@@ -1,17 +1,16 @@
 [Unit]
 Description=MicroOS Health Checker
 After=cloud-final.service
-After=crio.service
-After=etcd.service
-After=kubelet.service
 After=rebootmgr.service
 After=systemd-logind.service
 Wants=local-fs.target
+Before=boot-complete.target
```
Comment (Vogtinator) [inline systemd/health-checker.service]: IIRC not necessary, as `boot-complete.target` requires this

File: sbin/health-checker.in [hunk 15]
```diff
@@ -71,77 +58,77 @@ telem_send_record()
...
+    #   indeterminate: when an entry is neither good or bad, i.e.
+    #                  we are still trying to boot 3 times
+    status=$(systemd-bless-boot status)
+    # systemd-boot fill the EFI variables with the info we need
+    
+    # Get the entry depending on the current booted snapshot
+    current_snapshot=$(findmnt --output OPTIONS --noheadings / | sed -e 's|.*subvol=/@/.snapshots/\([0-9]\+\)/snapshot.*|\1|g')
+    # We do not want a file that does not exist
+    shopt -s nullglob
+    selected_entry=(/boot/efi/loader/entries/*[a-z]-"$current_snapshot"+[[:digit:]]-[[:digit:]].conf /boot/efi/loader/entries/*[a-z]-"$current_snapshot".conf)
```
Comment (Vogtinator) [inline sbin/health-checker.in]: This should ideally not hardcode `/boot/efi/loader/entries` or the format of entries.

Would be nice to use `bootctl list --json=pretty | jq -r '.[]|select(.isSelected).id'` here, but I guess that does not work with grub?

**Review by Vogtinator — CHANGES_REQUESTED** [review]
`grep` without `-q` pollutes stdout

File: README.md (hunk 4 above)
Comment (danyspin97) [reply README.md]: `health-checker` reads both `/usr/libexec/health-checker` and `/usr/local/libexec/health-checker`; though the default plugins are installed in the former. I think we can highlight that plugins can be installed in both? Though I don't really like adding files into `/usr/libexec` not managed by the package manager

File: README.md (hunk 5 above)
Comment (danyspin97) [reply README.md]: My bad, this should have been `/usr`.

File: README.md (hunk 6 above)
Comment (danyspin97) [reply README.md:35]: The boot count is enabled by `kernel-install` when `/etc/kernel/tries` is set, so this only works for new snapshot. `systemd-bless-boot` will reset the counter when the new snapshot boots successfully, and it is never enabled again.

File: systemd/health-checker.service (hunk 14 above)
Comment (danyspin97) [reply systemd/health-checker.service]: I'll remove it, it was included to mirror the example `/usr/lib/systemd/system/systemd-boot-check-no-failures.service`.

File: sbin/health-checker.in (hunk 15 above)
Comment (danyspin97) [reply sbin/health-checker.in]: You are right, `bootctl` has not been used because it doesn't work with grub. The current method isn't really accurate as `bootctl` should be used for getting systemd-boot default entry and grub2 saves the default into `grubenv` file.

Comment (aplanas) [reply sbin/health-checker.in]: there is a bli.mod in grub2 (https://www.gnu.org/software/grub/manual/grub/html_node/bli.html) but only populate two EFI variables. If this gets improved we can use bootctl

Comment (danyspin97) [reply sbin/health-checker.in]: I have a patch for grub2 that populates `LoaderBootCountPath` since it need by health-checker. I'll look up which other variables bootctl uses.

File: README.md (hunk 4 above)
Comment (thkukuk) [reply README.md]: Even if not snapshoted, we neeed `/usr/local/libexec/health-checker`, too. This was the clear feedback from the past. Not everybody wants to fiddle around with transactional-update to install a health-checker plugin which is not coming as RPM.
And yes, we should esplain that /usr/libexec/... is for plugins coming through an RPM and /usr/local/libexec/... for manual installation.

File: sbin/health-checker.in (hunk 15 above)
Comment (aplanas) [reply sbin/health-checker.in]: This would be the best solution. The variables are described here: https://systemd.io/BOOT_LOADER_INTERFACE/ and I think that .isSelected comes from LoaderEntrySelected. There are others that would help, for example LoaderEntryDefault or LoaderConfigTimeout

File: plugins/health-check-tester.sh [hunk 16]
```diff
@@ -6,6 +6,11 @@
 
 run_checks() {
 
+   # Check if the system is running legacy grub as the check file
+   # is only used there
+   if [ -z "$(bootctl list 2>/dev/null)" ]; then
+       return 0
+   fi
```
Comment (aplanas) [inline plugins/health-check-tester.sh:13]: To detect the bootloader I would suggest `sdbootutil bootloader`, or `update-bootloader --show` or even `grep LOADER_TYPE /etc/sysconfig/bootloader`.

Ideally `bootctl is-installed` should be used, but if we need to depend of the presence of a boot entry list, this can also work `[ -d /boot/efi/loader/entries]`

File: sbin/health-checker.in [hunk 17]
```diff
@@ -21,6 +21,11 @@ BTRFS_ID_DEFAULT=0
 SNAPSHOT_DEFAULT=""
 BTRFS_ID_CURRENT=0
 
+is_bls=1
+if [ -z "$(bootctl list 2>/dev/null)" ]; then
+    is_bls=0
+fi
```
Comment (aplanas) [inline sbin/health-checker.in:27]: same

File: plugins/health-check-tester.sh (hunk 16 above)
Comment (danyspin97) [reply plugins/health-check-tester.sh:13]: `bootctl is-installed` only checks for systemd-boot so it doesn't work for grub2-bls, unfortunately. I think `[ -d /boot/efi/loader/entries ]` might be the simplest solution.

Comment (aplanas) [reply plugins/health-check-tester.sh:13]: Maybe there is an EFI variable that determine if BLS is used, like for example `LoaderInfo` that is used by both bootloaders

Comment (danyspin97) [reply plugins/health-check-tester.sh:13]: From `bootctl is-installed` code:

```c
         /*  → It specifically checks for systemd-boot, not for other boot loaders (which a check for
         *    /boot/loader/entries would do). */
```

We can check for a specific EFI variable, but I would go for the simplest option here, considering that health-checker for bls relies on entry names and boot counting that wouldn't be available otherwise.

File: README.md [hunk 18]
```diff
@@ -9,10 +9,50 @@ Check the state of a openSUSE MicroOS system after a reboot.
...
+responsible to check a special service or condition. All services, which should
+be checked by the plugin, needs to be listed in the 'After' section. To run the
+check the plugin is called with the option *check*. If this fails, the plugin
+will exit with the return value `1`, else `0`. The `health-checker` script will
+call all plugins inside `/usr/libexec/health-checker` and
+`/usr/local/libexec/health-checker` directories. The former directory includes plugins
+installed by system packages (and therefore coming through an RPM), the latter includes
+plugins installed manually by the system admin. Every plugin is responsible to check
+a special service or condition. For this, the plugin is called with the option
+*check*. If this fails, the plugin will exit with the return value `1`, else
```
Comment (thkukuk) [inline README.md]: The check option is already mentioned some lines above, please merge this.

File: README.md [hunk 19]
```diff
@@ -9,10 +9,50 @@ Check the state of a openSUSE MicroOS system after a reboot.
...
+`0`. Have a look at the default plugins shipped in
+`/usr/libexec/health-checker` for examples.
+
+Its behavior depends if the system is using systemd-boot/grub2-bls (i.e.
```
Comment (thkukuk) [inline README.md]: Better: "is using systemd-boot, grub2-bls or any other bootloader following the Boot Loader Specification (BLS) or legacy..."
I wouldn't mention bootloader internal things like the /boot/efi/loader/entries, this can change in the future and a change is already under discussion.

File: README.md [hunk 20]
```diff
@@ -9,10 +9,50 @@ Check the state of a openSUSE MicroOS system after a reboot.
...
+assess the status of the system at the boot and act accordingly. It uses the
+[Automatic Boot Assessment](https://systemd.io/AUTOMATIC_BOOT_ASSESSMENT/)
+provided by systemd to check every new snapshot, to reboot a predetermined
+amount of times, and then it lets the user access an emergency shell when
+everything else fails.
+
+Every new snapshot has a separate boot entry with a boot counter (according to
+`/etc/kernel/tries`, which health-checker sets to 3 by default); when that
+snapshot is booted for the first time, the bootloader (systemd-boot by
+default on MicroOS, but grub2-bls is also supported) will decrease the amount
```
Comment (thkukuk) [inline README.md]: I would remove "(systemd-boot by default on MicroOS, but grub2-bls is also supported)", not relevant internal implementation detail which can change every time. And then we forget to adjust the text here.

File: man/health-checker.8.xml.in [hunk 21]
```diff
@@ -62,11 +62,20 @@
...
-  rebooted successfully before, a reboot is tried. If this does not help,
-  some sevices are shutdown and an admin has to repair the system.
+  The behavior depends if the system is using systemd-boot/grub2-bls
+  or legacy grub2.
+  When the former is used, health-checker checks if it is the first time
+  a snapshot is booted or not. If it is the first time, the systemd will reboot a
+  number of timed configured in <filename>/etc/kernel/tries</filename>. If the
+  system still isn't working, then an emergency shell is started. If it is not
+  the first boot with the selected snapshot, then an emergency shell is
+  automatically started.
```
Comment (thkukuk) [inline man/health-checker.8.xml.in]: Does this mean we will never do an automated rollback with systemd-boot or grub2-bls?

File: sbin/health-checker.in [hunk 22]
```diff
@@ -133,15 +155,76 @@ error_decission()
...
+    status=$(systemd-bless-boot status)
+    # The bootloader fills the EFI variables with the info we need
+    
+    # Get the booted entry
+    current_entry=$(bootctl list --json=short | jq ".[] | select(.isSelected) | .id")
+    should_reboot=0
+    # Do not reboot by default if the entry has been chosen manually or the reboot has
+    # been disabled in the kernel cmdline
+    # selected_entry contains the boot count, remove it before comparing it to the default entry
+    if ! grep -qw "health-checker-reboot=disabled" /proc/cmdline; then
```
Comment (thkukuk) [inline sbin/health-checker.in:204]: Please document this option in the manual page.

**Review by thkukuk — CHANGES_REQUESTED** [review]

File: README.md (hunk 18 above)
Comment (danyspin97) [reply README.md]: Fixed.

File: README.md (hunk 19 above)
Comment (danyspin97) [reply README.md]: > I wouldn't mention bootloader internal things like the /boot/efi/loader/entries, this can change in the future and a change is already under discussion.

Thanks for pointing it out Thorsten, I wasn't aware of this.

File: README.md (hunk 20 above)
Comment (danyspin97) [reply README.md]: Fixed.

File: man/health-checker.8.xml.in (hunk 21 above)
Comment (danyspin97) [reply man/health-checker.8.xml.in]: The description doesn't reflect the current behavior of health-checker. I have updated the description.

> Does this mean we will never do an automated rollback with systemd-boot or grub2-bls?

The rollback is handled by rebooting and choosing a different boot entry.

File: sbin/health-checker.in (hunk 22 above)
Comment (danyspin97) [reply sbin/health-checker.in:204]: Fixed.

File: README.md [hunk 23]
```diff
@@ -8,11 +8,47 @@ Check the state of a openSUSE MicroOS system after a reboot.
...
+amount of times, and then it lets the user access an emergency shell when
+everything else fails.
+
+Every new snapshot has a separate boot entry with a boot counter (according to
+`/etc/kernel/tries`, which health-checker sets to 3 by default); when that
+snapshot is booted for the first time, the bootloader will decrease the amount
+of tries left. Then if health-checker succeed, systemd will call
+`systemd-bless-boot`, which will mark the new snapshot as working. If instead
+`health-checker` fails, it will reboot the current entry until the amount of
+tries available equals 0. At that point an emergency shell gets opened.
```
Comment (laenion) [inline README.md]: In contrast to the man page the _README_ doesn't mention rollbacks with BLS based systems at all, it just mentions the counter and the emergency shell. Which one is correct?

File: man/health-checker.8.xml.in [hunk 24]
```diff
@@ -62,11 +62,27 @@
...
-  to the last known working snapshot is executed. If the snapshot was already
-  rebooted successfully before, a reboot is tried. If this does not help,
-  some sevices are shutdown and an admin has to repair the system.
+  The behavior depends if the system is using systemd-boot/grub2-bls
+  or legacy grub2.</para>
+  <para>With systemd-boot/grub2-bls: the bootloader picks the first available entry
+  and health-checker checks that the system has booted successfully. When
+  any plugin fails, then health-checker will behave depending on the entry
+  booted: if it is a snapshot that has previously been booted successfully
+  than health-checker will try rebooting the system once more, and it will
```
Comment (laenion) [inline man/health-checker.8.xml.in]: "than" -> "then"

File: sbin/health-checker.in [hunk 25]
```diff
@@ -21,13 +21,29 @@ BTRFS_ID_DEFAULT=0
...
+get_snapshot()
+{
+    sed -e 's|.*@/\.snapshots/\([0-9]\+\)/snapshot.*|\1|g'
+}
+
+set_snapshot()
+{
+    SNAPSHOT_DEFAULT="$(btrfs subvolume get-default / | get_snapshot)"
+    SNAPSHOT_CURRENT="$(findmnt --output OPTIONS --noheadings / | get_snapshot)"
+}
```
Comment (laenion) [inline sbin/health-checker.in]: This function doesn't set the snapshot as the name implies - it's setting two variables containing snapshots.

File: sbin/health-checker.in [hunk 26]
```diff
@@ -21,13 +21,29 @@ BTRFS_ID_DEFAULT=0
...
 
+get_snapshot()
+{
+    sed -e 's|.*@/\.snapshots/\([0-9]\+\)/snapshot.*|\1|g'
+}
+
+set_snapshot()
+{
+    SNAPSHOT_DEFAULT="$(btrfs subvolume get-default / | get_snapshot)"
+    SNAPSHOT_CURRENT="$(findmnt --output OPTIONS --noheadings / | get_snapshot)"
```
Comment (laenion) [inline sbin/health-checker.in]: It can happen that a `transactional-update apply` was performed before the health check is running (e.g. during initrd as done by some of our cloud images). In this case the wrong snapshot will be put into the variable here, so this would also need the check from https://github.com/openSUSE/transactional-update/blob/v5.5.0/sbin/transactional-update.in#L948-L951

**Review by laenion — COMMENTED** [review]
Just a few minor finds and one actual bug :-)

File: sbin/health-checker.in (hunk 25 above)
Comment (danyspin97) [reply sbin/health-checker.in]: Right, the name it's misleading. I renamed it to `set_snapshot_id`, to be consistent with the rest of the script.

File: sbin/health-checker.in (hunk 26 above)
Comment (danyspin97) [reply sbin/health-checker.in]: I wasn't aware of this use case, thanks for pointing out the fix!

File: README.md (hunk 23 above)
Comment (danyspin97) [reply README.md]: I have updated the description in the README. Booting a different snapshot is equivalent of doing a rollback, but this wasn't clearly explained in the README.
